Implement ARRAY_AGG(DISTINCT ...)

james727 opened the pull request:
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/1323

# What changes are included in this PR?
This includes the implementation of `array_agg(distinct ...)`. I also moved `phsyical_plan/distinct_expression.rs` into `physical_plan/expression` - I wanted to reuse a private test utility in `physical_plan/expression`, and it seemed like a more natural location for it anyways. I can move it back if preferred though.

# Are there any user-facing changes?
This adds net-new functionality. Not sure if any doc updates are needed though.

# Input needed
I'm interested in any tips on improving the testing approach. It's pretty annoying that the output order is non-deterministic here, and the lack of `ORD` implementation for `ScalarValue` makes a sort + equality check approach difficult as well. I'm new to Rust though so I'm sure there are better ways to structure this than what I've done so far.

alamb commented on datafusion/src/physical_plan/aggregates.rs:
```diff
@@ -454,42 +452,40 @@ mod tests {
                         assert_eq!(
                             Field::new("c1", DataType::UInt64, true),
-                            result_agg_phy_exprs_distinct.field().unwrap()
+                            result_distinct.field().unwrap()
```
well that is certainly much nicer to read 👍

alamb commented on datafusion/src/physical_plan/expressions/distinct_expressions.rs:
```diff
@@ -207,17 +208,155 @@ impl Accumulator for DistinctCountAccumulator {
+        )])
+    }
+
+    fn update(&mut self, _values: &[ScalarValue]) -> Result<()> {
```
I think this will need to be removed now that https://github.com/apache/arrow-datafusion/pull/1582 is merged

alamb commented on datafusion/src/physical_plan/expressions/distinct_expressions.rs:
```diff
@@ -207,17 +208,155 @@ impl Accumulator for DistinctCountAccumulator {
+        Ok(())
+    }
+
+    fn merge(&mut self, _states: &[ScalarValue]) -> Result<()> {
```
likewise this

alamb commented on datafusion/src/physical_plan/expressions/distinct_expressions.rs:
```diff
@@ -705,4 +844,151 @@ mod tests {
+
+        match (expected, actual) {
+            (ScalarValue::List(Some(e), _), ScalarValue::List(Some(a), _)) => {
+                // Check that the inputs are the same length.
+                assert_eq!(e.len(), a.len());
+
+                let h1: HashSet<ScalarValue> = HashSet::from_iter(e.clone().into_iter());
+                let h2: HashSet<ScalarValue> = HashSet::from_iter(a.into_iter());
+
+                // Check that e's elements are unique.
+                assert_eq!(h1.len(), e.len());
+
+                // Check that a contains the same unique elements as e.
+                assert_eq!(h1, h2);
```
ScalarValue does implement `PartialOrd`-- the reason it doesn't implement Ord is that it doesn't really make sense to compare different variant types -- e.g. `ScalarValue::List` and `ScalarValue::Boolean`

However, in your test cases you are always comparing the same enum, so you can do something like the following (which will panic if there is a problem comparing the `ScalarValue`s in the test):


```suggestion
                // workaround lack of Ord of ScalarValue
                let cmp = |a: &ScalarValue, b: &ScalarValue| {
                    a.partial_cmp(b).expect("Can compare ScalarValues")
                };
                e.as_mut().sort_by(cmp);
                a.as_mut().sort_by(cmp);

                // Check that the inputs are the same
                assert_eq!(e, a);
```

james727 replied:
Thank you! This is much nicer - I noticed the `PartialOrd` implementation but was unsure of how to actually use it.

alamb commented on datafusion/tests/sql/aggregates.rs:
```diff
@@ -410,3 +411,58 @@ async fn csv_query_array_agg_one() -> Result<()> {
+    let list = ScalarValue::try_from_array(column, 0)?;
+    let mut contents = vec![];
+    match list {
+        // Since ScalarValue does not implement ORD, we must extract the underlying values
```
👍

alamb reviewed (approved):
@james727  I am sorry for the delayed review. There has been a lot going on in arrow and datafusion recently!

The code looks (very) good to me; However, there will be a logical conflict when it is merged with the latest master (need to remove `update` and `merge`). Can you please do so and I'll get this PR merged?

I also left a suggestion about how you can work around the lack of `Ord` in `ScalarValue`

alamb commented:
There are a few conflicts in this PR that need to be cleaned up before it can be merged.

james727 commented:
No worries @alamb - I've noticed a ton of activity and figured this is a relatively low-priority change. Appreciate the review. I've rebased to fix conflicts and cleaned up the tests with your suggestion, so this is ready for another look whenever you have time.

Re: conflicts - yep, sorry for the confusion. I noticed your review last night and responded to a couple of things but didn't have time to update the PR.

alamb commented:
Thanks @james727 !
